File: 백준/Gold/9663. N－Queen/N－Queen.py
import sys
myinput = sys.stdin.readline

N= int(myinput())
col = [0] * (N)
answer = 0

def isokay(i):
    k = 0
    while k<i:
        if abs(col[i]-col[k])==abs(i-k) or (col[i]==col[k]):
            return False
        k+=1
    return True

def finding(i):
    global answer
    if i==N:
        answer+=1
    else:
        for j in range(N):
            col[i] = j
            if isokay(i):
                finding(i+1)



finding(0)
print(answer)

File: 백준/Gold/9663. N－Queen/N－Queen.cc
#include <iostream>
#define MAX 15
using namespace std;

int col[MAX];
int N, answer=0;


bool isokay(int i){
    int k=0;
    while (k<i){
        if (col[i]==col[k] || abs(col[i]-col[k])==i-k){
            return false;
        }
        k++;
    }
    return true;
}

void finding(int i){
    if (i==N){
        answer ++;
    } else {
        for (int j=0; j<N; j++) {
            col[i] = j;
            if (isokay(i))
                finding(i+1);
        }
    }
}

int main() {
    cin >> N;
    finding(0);
    cout<<answer;
    
}